Import Forward Declared Objective-C Interfaces and Protocols

NuriAmari opened the pull request:
This patch modifies the ClangImporter to synthesize a placeholder for forward declared Objective-C interfaces and Protocols.

This is the implementation for [SE-0384](https://github.com/apple/swift-evolution/blob/main/proposals/0384-importing-forward-declared-objc-interfaces-and-protocols.md)

NuriAmari commented:
@swift-ci Please smoke test

drodriguez commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented on lib/Serialization/Serialization.cpp:
```diff
@@ -619,9 +619,9 @@ DeclID Serializer::addDeclRef(const Decl *D, bool allowTypeAliasXRef) {
           isa<PrecedenceGroupDecl>(D)) &&
          "cannot cross-reference this decl");
 
-  assert((!D || !isDeclXRef(D) ||
```
Eventually this needs addressing. The assert was introduced by https://github.com/apple/swift/pull/17223, which as I understand it was designed to combat issues with ambiguous protocol conformances in SourceKit (ie. a protocol and forward declared interface share a name). 

The assert is tripped during serialization of SILModule that references the forward declared type from another module. In this case, the declaration exists, is "cross reference", and has the `ForbidSerializationReferenceAttr` attribute. Perhaps these declarations need not be included in the serialized result, but I'm not sure of the implications. If such a workaround is not possible, perhaps a diagnostic suggesting a source level fix is appropriate.

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci please test

NuriAmari commented:
@swift-ci Please Test Source Compatibility

NuriAmari commented:
@swift-ci please test

NuriAmari commented:
@swift-ci please test source compatibility

NuriAmari commented:
@swift-ci please smoke test

NuriAmari commented:
@swift-ci please test source compatibility

NuriAmari commented:
@swift-ci please stress test

NuriAmari commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1273,6 +1275,12 @@ static bool ParseClangImporterArgs(ClangImporterOptions &Opts,
   Opts.DumpClangDiagnostics |= Args.hasArg(OPT_dump_clang_diagnostics);
 
+  if (FrontendOpts.RequestedAction != FrontendOptions::ActionType::REPL &&
+      FrontendOpts.RequestedAction != FrontendOptions::ActionType::NoneAction &&
```
When invoked directly (ie. `lldb --repl="...") the `RequestedAction` seems to be `NoneAction`.

beccadax replied:
Should this be a comment in the code?

NuriAmari commented on test/ClangImporter/incomplete_objc_types_importing_limitations_diagnostics.swift:
```diff
@@ -0,0 +1,25 @@
+incompleteInterface.doSomethingForwardDeclaredInterfacesCan()
+// CHECK: incomplete_objc_types_importing_limitations_diagnostics.swift:{{[0-9]+}}:{{[0-9]+}}: error: value of type 'ForwardDeclaredInterface' has no member 'doSomethingForwardDeclaredInterfacesCan'
+// CHECK: incompleteInterface.doSomethingForwardDeclaredInterfacesCan()
+// CHECK: <unknown>:0: note: class 'ForwardDeclaredInterface' is a placeholder for a forward declared Objective-C interface and may be missing members; import the definition to access the complete interface
```
`<unknown>:0` looks a little unfortunate, but there is no source location available for a synthesized type.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@swift-ci please stress test

NuriAmari commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4322,17 +4316,51 @@ namespace {
+              importingContext.getValue()->getParentSourceFile();
+
+          // TODO: Make sure we aren't bailing early in legitimate cases
+          if (!importingSwiftSourceFile)
```
Double check

NuriAmari commented on lib/AST/Module.cpp:
```diff
@@ -1024,6 +1049,53 @@ void ModuleDecl::getTopLevelDeclsWhereAttributesMatch(
+        const auto &declObjCAttribute = VD->getAttrs().getAttribute<ObjCAttr>();
+        const Identifier &declObjCName =
+            declObjCAttribute->hasName()
+                ? declObjCAttribute->getName()->getSimpleName()
```
Are there no top level decls with non simple names? Existing implementation seems to imply not, but I wasn't sure: https://github.com/apple/swift/blob/main/lib/ClangImporter/ImportDecl.cpp#L4282

NuriAmari replied:
> So, can topLevelObjCExposedDeclsInFileUnit contain global functions? If so, then you might see compound names here.

I don't see why not, I'll write a test to check.

NuriAmari replied:
> So, can topLevelObjCExposedDeclsInFileUnit contain global functions? If so, then you might see compound names here.

Upon further investigation -- no. Only declarations with an @objc attribute appear in that list and as it turns out, you can't @objc annotate a Swift function.

```
test.swift:1:2: error: @objc can only be used with members of classes, @objc protocols, and concrete extensions of classes
@objc public func compoundNameTopLevelDecl(foo foo: Int, bar bar: Int) {
~^~~~~
```
So the declarations Swift name has to be simple. I also tried if it is for some reason possible to provide a Swift declaration with a compound @objc name, but that fortunately also isn't allowed:

```
test.swift:3:7: error: '@objc' class must have a simple name
@objc(my:compound:name:) public class Foo : NSObject {
      ^ ~~~~~~~~~~~~~~~
test.swift:9:7: error: '@objc' protocol must have a simple name
@objc(my:other:compound:name:) public protocol Bar {}
      ^ ~~~~~~~~~~~~~~~~~~~~~
test.swift:12:7: error: '@objc' extension must have a simple name
@objc(my:last:compound:name:) public extension Foo {
      ^ ~~~~~~~~~~~~~~~~~~~~
```

NuriAmari commented on lib/AST/Module.cpp:
```diff
@@ -1024,6 +1049,53 @@ void ModuleDecl::getTopLevelDeclsWhereAttributesMatch(
+  if (!isObjCNameLookupCachePopulated())
+    populateObjCNameLookupCache();
+
+  if (name.isSpecial())
```
Are we dropping anything here? I modeled this after the SwiftLookupTable, which also seems to bail on "special" names: https://github.com/apple/swift/blob/main/lib/ClangImporter/SwiftLookupTable.h#L69-L71.

NuriAmari replied:
I suppose if you wrote `@class init`, `@class dealloc` or `@class subscript`, maybe? Does merely use the same spelling make them special? I somehow doubt it, but I will check.

NuriAmari replied:
Added it to a test, seems fine.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4293,12 +4287,12 @@ namespace {
-        // This shouldn't be allowed but we need it for source compatibility;
-        // people used `@class SwiftNameOfClass` as a workaround for not
-        // having the previous loop, and it "worked".
+      if (allowObjCMismatchFallback && !found) {
```
Should be removed all together in another PR for swift language mode 6+

NuriAmari commented:
@swift-ci Please test :)

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@swift-ci Please stress test

NuriAmari commented:
This is the implementation for [SE-0384](https://forums.swift.org/t/se-0384-importing-forward-declared-objective-c-interfaces-and-protocols/62392/4). I've left some comments in areas I'm not sure about that are worth taking a look, sorry about all the ci comments.

**Changes:**

During the proposal review it was discussed briefly if having the placeholder protocol inherit from `NSObjectProtocol` was sound. Seems there are legitimate use cases of importing non `NSObjectProtocol` inheriting protocols, and the benefit seems limited, so I have removed it. I've added a test for this, as well as a test for NSProxy, which was also raised during the review.

I've also added resolution of forward declared classes and protocols to their `@objc` exposed Swift definitions. This already happened automatically with Clang definitions, we need this to make the experience consistent with Swift definitions.

**Testing:**

I have this draft PR https://github.com/apple/swift/pull/62649 which is identical to this one, save for the flag is enabled by default. I used this to test against the source compatibility suite and source kit stress test. The release variant of source compatibility is failing at the moment, but its just broken for unrelated reasons in my estimation.

I also have this LLDB draft PR where I am writing tests to ensure importing forward declarations is gracefully disabled in the REPL and during expression evaluation in LLDB: https://github.com/apple/llvm-project/pull/6114. I'm still working out how to write the LLDB test, but looks fine so far.

NuriAmari commented:
Hey, @zoecarver @beccadax could I get some feedback on this 🙂 ?. Happy to ask anyone else, not sure who owns ClangImporter related changes.

beccadax commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4322,17 +4316,50 @@ namespace {
             return result;
         }
       }
+      if (Impl.ImportForwardDeclarations) {
```
I think I'm missing something important here: *Why* are we doing this? I can see that we're looking in the use sites imported modules for the name we're trying to import, and that we're only doing it when `ImportForwardDeclarations` is enabled, but I don't understand the overall purpose. Why do we need to look in the imported modules?

I'm basically asking because looking in the imported modules of the use sites seems *really, really weird* and I'm not sure it will interact well with the way ClangImporter caches declarations. (After all, couldn't we try to import the same declaration from use sites with different imported modules? What happens then?)

NuriAmari replied:
So this `resolveSwiftDecl` function, even prior to this patch was used only to resolved forward declared interfaces and protocols in Objective-C headers to a Swift definition. As I understand it, before this patch, it only looked in the overlay module to support mixed modules.

**Why are we now also looking at the site's imported modules?**

Well to start, to match the behavior of Clang modules. When you import Clang modules A and B into a Swift file, a forward declaration `@class Foo` in module A, will be resolved to a definition `@interface Foo ...` in module B. When you use the same `clang::CompilerInstance` to load two modules, the AST is "merged". But if the definition is instead in **Swift** module C, this won't work anymore.

This is really confusing for users not versed in Clang modules vs Swift modules. Perhaps misguided, but the conventional advice for dealing with forward declarations without bloating your ObjC headers with unnecessary includes has been to "import the defining module". 

Specifically with reference to this patch, not doing this leaves us vulnerable to other problems. You can see an example in `incomplete_objc_types_swift_definition_imported.swift`. If we start importing placeholder types for forward declarations, _without_ searching for a visible Swift @objc definition, we can end up with two conflicting Swift representations for the same type. 

**Why are we only doing it when ImportForwardDeclarations is enabled?**

As mentioned, it is required for importing forward declarations safely and source breaking so needs to be gated behind the flag. It would no doubt be useful separate from this feature, but that seems like a different patch.

**Isn't it an issue if different use sites have different imported modules?**

I don't think so. We are only looking to resolve the name to an @objc annotated Swift declaration, nothing else. To my knowledge, @objc provided names are not "namespaced" by the defining module. I don't think `@objc(Foo) class Bar {}` and `@objc(Foo) class Baz {}` in the same program is supported, regardless of which module they are defined in (I've seen warnings along the lines of "multiple selectors for Foo, which implementation is used is undefined"). All that to say if we find one definition, that should be the only one. Similarly once one definition 
 is cached, it should be fine to continue returning that. That said, I could be wrong about the rules around @objc name conflicts, so let me know if I've missed something.

NuriAmari replied:
EDIT: I've uncovered problems, need to take another run at it.

beccadax replied:
> If we start importing placeholder types for forward declarations, without searching for a visible Swift @objc definition first, we can end up with two conflicting Swift representations for the same type (one directly from the imported Swift module, the other created to represent a forward declaration of the same Swift type).

To be clear, if we *do* find a type this way, is the API using it still imported as unavailable? Or does it mean that APIs that were previously dropped because they relied on a non-modular forward declaration are now imported *and usable*? The latter doesn't seem to be within the scope of the Evolution proposal, and I have concerns about whether it might end up producing corrupt module files (because some of their declarations use types from modules that aren't in the module's import list).

> If it helps, we could also look at all Swift modules imported into the entire main module? I don't really like that though.

I think I would prefer that because it would mean ClangImporter was always considering the same set of imported modules, at least within a given compilation job.

NuriAmari replied:
> To be clear, if we do find a type this way, is the API using it still imported as unavailable? Or does it mean that APIs that were previously dropped because they relied on a non-modular forward declaration are now imported and usable? 

I'm not sure what you mean by a non-modular forward declaration, but yes the API relying on it would then be imported and usable. 

> The latter doesn't seem to be within the scope of the Evolution proposal, and I have concerns about whether it might end up producing corrupt module files (because some of their declarations use types from modules that aren't in the module's import list).

Fair enough, admittedly this problem didn't occur to me until later in the evolution process. If that's the case, I will have to start another round of evolution as I don't think this change is feasible without this.

Could you elaborate a little on the module's import list? I'm not familiar. In what scenario would a module be available while compiling a module, but not included in the produced module's import list?

> I think I would prefer that because it would mean ClangImporter was always considering the same set of imported modules, at least within a given compilation job.

Yeah I've made this change, I don't think we have any other choice. Could you elaborate on the compilation job comment? I was investigating a bit if the imported modules might change between each front-end invocation of an incremental build, but in the testing I did, it seemed ok? Am I wrong in this belief?

NuriAmari commented on include/swift/AST/Decl.h:
```diff
@@ -629,54 +629,61 @@ class alignas(1 << DeclAlignInBits) Decl : public ASTAllocated<Decl> {
+      // If the map from @objc provided name to top level swift::Decl in this
+      // module is populated
+      ObjCNameLookupCachePopulated : 1);
 
```
Sure. For future reference, should I be manually formatting? This formatting change is `clang-format`'s doing.

NuriAmari commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -3315,7 +3321,13 @@ void ClangModuleUnit::lookupValue(DeclName name, NLKind lookupKind,
   if (auto lookupTable = owner.findLookupTable(clangModule)) {
     // Search it.
-    owner.lookupValue(*lookupTable, name, *consumer);
+    if (importingModuleOrFile) {
+      owner.setImportingContext(*importingModuleOrFile);
+      owner.lookupValue(*lookupTable, name, *consumer);
+      owner.resetImportingContext();
+    } else {
+      owner.lookupValue(*lookupTable, name, *consumer);
+    }
```
> Okay, I see, you're using it deep inside ImportDecl and threading the information through to there might be tricky.

Yeah I tried this, but IIRC I gave up trying to thread it down when I hit the `SwiftDeclConverter` visitor. Not the first time I've run into this.

NuriAmari replied:
No more context required.

NuriAmari commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -6195,7 +6207,7 @@ void ClangImporter::diagnoseTopLevelValue(const DeclName &name) {
 void ClangImporter::diagnoseMemberValue(const DeclName &name,
                                         const Type &baseType) {
-  if (!baseType->getAnyNominal())
+  if (!(baseType->getAnyNominal() || baseType->is<ExistentialType>()))
```
Unrelated bug fix I suppose. We could always get an `ExistentialType` here, but if we tried to diagnose a member of one before now, it wasn't working. I didn't realize this until I wrote: `incomplete_objc_types_importing_limitations_diagnostics.swift`.

When I added this early return originally, the motivation was to bail on types that `namelookup::extractDirectlyReferencedNominalTypes` would throw on: https://github.com/apple/swift/pull/41079#discussion_r795089560. Maybe best to expand the early return to match all types the function can handle? In a separate patch?

NuriAmari replied:
I've expanded the early return check to allow all types that `namelookup::extractDirectlyReferencedNominalTypes` knows how to handle. Let me know if I should move that into another commit or PR.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
**Updates:**

- When attempting to resolve a forward declared type to a definition, look at all imported Swift modules, not just those imported at the use site. This is required for the scenario shown in test: `incomplete_objc_types_compatibility_complete_incomplete_inter_file_swift_definition.swift`.
    - It turns out Clang modules also behave this way (ie. importing the defining module makes resolves forward declarations, regardless of what file they are imported in)
- Since we no longer need to know the imports visible at the use site, I've removed the `lookUpValueWithContext` related code.

**Potential Improvement:**

We no longer care what module an @objc annotated Swift declaration comes from, as long as it is available. Our lookup cache no longer needs to be module specific, we could save some time by not iterating through them all every time and just doing the lookup on the cache directly.

allevato commented:
Hi @NuriAmari , we had a chance to discuss this with the full language workgroup, and we came to the decision that supporting forward declarations to `@objc class`es defined in Swift would require a separate pitch/proposal. Even though Clang allows it, Swift's lack of separation between interface and implementation prevents the kind of dependency cycles that forward declarations can be used to break, and we don't want Obj-C forward declarations to act as a "back door" to create these kinds of semi-cyclic dependencies in Swift.

For SE-0384 and this PR, we'd like to have the compiler emit an error if a Swift type is found that matches a forward-declared type but it is *not* defined in the same module (via an overlay or bridging header). Can you make that change?

NuriAmari commented:
> For [SE-0384](https://apple.github.io/swift-evolution/#?proposal=SE-0384) and this PR, we'd like to have the compiler emit an error if a Swift type is found that matches a forward-declared type but it is _not_ defined in the same module (via an overlay or bridging header). Can you make that change?

Just to be clear, emit an error and still synthesize a placeholder definition for the forward declaration, or emit an error and drop the forward declaration? Personally until another pitch / proposal can go through, I prefer the latter (dropping).

With regards to the separate pitch / proposal I've created the pitch: https://forums.swift.org/t/pitch-resolve-objective-c-forward-declarations-to-objc-swift-definitions/62920. If the language group has specific concerns, I'd appreciate feedback as I'll likely pursue that after this lands.

allevato commented:
> If the user attempts to use a forward declaration that is found to be defined in a different Swift module, we attach a note to the forward declaration. We do _not_ synthesize a placeholder. The user referencing any declaration that depends upon this forward declaration, will see the note, pointing out the forward declaration and explaining that it "shadows" an @objc Swift definition in another module, and was thus not imported. If possible, I will point to that Swift definition as well.
> 
> Drawbacks to consider. Importing new Swift modules will then have the ability to _break_ previously working code. Maybe justly so, but I'm sure users will find this confusing. Let me know if that sounds reasonable.

That sounds reasonable, but if you're only referring to notes, where/on which one do you intend to emit the error about the shadowing?

Regarding the last point, I don't think that breakage would occur in practice because forward-declaring a Swift `@objc class` already shouldn't be expected to work. As an example, let's say Swift module `A` defines an `@objc class`, and Clang module `B` only forward declares that class and has some API that references it. Even if Swift module `C` that imports both `A` and `B`, the APIs in `B` that refer to `A` won't work, so anyone trying to forward-declare a Swift class in Obj-C is already putting their build in a fragile and unreliable state.

NuriAmari commented:
> That sounds reasonable, but if you're only referring to notes, where/on which one do you intend to emit the error about the shadowing?

So the current diagnostics look like this:

```
_ = giveAFoo() // This doesn't!

test.swift:5:5: error: cannot find 'giveAFoo' in scope
_ = giveAFoo()
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
FooObjCWrapper.h:3:1: note: function 'giveAFoo' not imported
Foo* giveAFoo();
^
FooObjCWrapper.h:3:1: note: return type not imported
Foo* giveAFoo();
^
FooObjCWrapper.h:3:1: note: interface 'Foo' is incomplete
Foo* giveAFoo();
^
FooObjCWrapper.h:3:1: note: interface 'Foo' forward declared here
@class Foo;
^
```

I was thinking the "interface 'Foo' is incomplete" would be replaced with something about being incomplete and shadowing a Swift declaration. Then ideally I'd point at the Swift declaration, like I do the forward declaration, but I've never tried to do something like that, so I can't make any promises.

So the answer to "where / on which one" is, the one you are "using". The one in scope from the Clang declaration you are using, but that can't be imported. To be clear here, the note _is_ the error about shadowing. 

> so anyone trying to forward-declare a Swift class in Obj-C is already putting their build in a fragile and unreliable state.

Fair enough, I don't think that was clear to people.

allevato commented:
I think that sounds reasonable; @beccadax do you have any concerns here?

NuriAmari commented:
@swift-ci Please test

NuriAmari committed:
Implement importing of forward declared objc protocols and interfaces

This modifies the ClangImporter to introduce an opaque placeholder
representation for forward declared Objective-C interfaces and
protocols when imported into Swift.

In the compiler, the new functionality is hidden behind a frontend
flag -enable-import-objc-forward-declarations, and is on by default
for language mode >6.

The feature is disabled entirely in LLDB expression evaluation / Swift
REPL, regardless of language version.

NuriAmari commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4347,27 +4349,81 @@ namespace {
+        moduleImportFilter |= ModuleDecl::ImportFilterKind::ImplementationOnly;
+        moduleImportFilter |= ModuleDecl::ImportFilterKind::SPIAccessControl;
+        moduleImportFilter |= ModuleDecl::ImportFilterKind::SPIOnly;
+        moduleImportFilter |= ModuleDecl::ImportFilterKind::ShadowedByCrossImportOverlay;
```
I don't know what kind of import requires the `ShadowedByCrosssImportOverlay` filter, but I figured I should include it.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4414,6 +4463,39 @@ namespace {
             decl->getSourceRange().getBegin());
 
+        if (Impl.ImportForwardDeclarations) {
+          if (auto native = hasNonLocalNativeSwiftDecl<ProtocolDecl>(decl, name, hasKnownSwiftName)) {
```
I tried to produce a diagnostic that points at the location of `native`, but the source location on the declaration is invalid. Instead I've added the module that declaration is found in to the new diagnostic note.

If the source location issue is resolved, we can add a better diagnostic.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
Hi @allevato @beccadax this is ready for review again. As described I have stopped synthesizing a placeholder if a matching Swift declaration is found and added a new diagnostic:

```
objc-library-forward-declaring-complete-swift-types.h:X:Y: note: incomplete interface 'Foo' shadows an imported @objc Swift class from module 'CompleteSwiftTypes'; import the generated Swift header for module 'CompleteSwiftTypes' in this header to realize this declaration
@class Foo;
^
```

If its a protocol, interface / class are replaced with protocol but the diagnostic is otherwise the same for protocols. I tried adding a note to point out the "offending" @objc Swift declaration, but the source location on these declarations is invalid for some reason (see comment above).

CI has been quite unstable recently, so I'm not sure it will pass, but everything as of late has looked unrelated. Please let me know if there's anything that needs adjusting 🙂 . Thanks!

NuriAmari commented on test/ClangImporter/incomplete_objc_types_swift_definition_imported.swift:
```diff
@@ -0,0 +1,173 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -parse-as-library %S/Inputs/custom-modules/IncompleteTypes/complete-swift-types.swift -module-name CompleteSwiftTypes -emit-module -emit-module-path %t/CompleteSwiftTypes.swiftmodule
+// RUN: not %target-swift-frontend -enable-import-objc-forward-declarations -enable-objc-interop -typecheck -I %S/Inputs/custom-modules/IncompleteTypes -I %t %s 2>&1 | tee ~/two-one.txt | %FileCheck %s
```
Will remove `tee` invocations 🙂

beccadax commented:
> I was thinking the "interface 'Foo' is incomplete" would be replaced with something about being incomplete and shadowing a Swift declaration. Then ideally I'd point at the Swift declaration, like I do the forward declaration, but I've never tried to do something like that, so I can't make any promises.

The terminology isn't quite right here: decl A shadows decl B if A is "closer" during name lookup and therefore is the declaration that the compiler selects, so here, the complete declaration is shadowing the incomplete one, not the other way around.

I might avoid the idea of "shadowing" and instead say something like:

```
interface 'Foo' is incomplete and cannot be imported as a stub; its name conflicts with a %0 in module %1
```

Where %0 is the `DescriptiveDeclKind` of the complete Swift declaration, and %1 is the module that contains the complete Swift declaration.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@beccadax When you get a chance, I've updated the PR with the diagnostic you suggested.

NuriAmari commented:
Source compat passes with the flag on: https://github.com/apple/swift/pull/62649. Stress test fails, but other jobs are failing in the same way, seems unrelated. The Windows test here isn't really pending, it finished successfully: https://ci-external.swift.org/job/swift-PR-windows/8806/

NuriAmari commented:
@swift-ci Please test windows platform

NuriAmari commented:
Hi everyone,

I wanted to draw attention to two aspects of this implementation that maybe were not entirely clear.

1. The ClangImporter option `ImportForwardDeclarations` existed before this change and was used by `swift-ide-test` to print interfaces with placeholders to represent forward declared interfaces and protocols. Per the discussion above, we are no longer synthesizing a placeholder if the declaration corresponds to a Swift definition. This also goes for `swift-ide-test` and will have an immediate effect for anyone using `swift-ide-test` and passing `--enable-objc-forward-declarations`. I think `swift-ide-test` and the compiler should be as consistent with one another as possible, so this makes sense, but I wanted to make sure to call out it is possible we break something here, but only for clients forward declaring Swift types.

2. I previously voiced some concern about this implementation allowing users to break previously working code by importing a new Swift module:

> Drawbacks to consider. Importing new Swift modules will then have the ability to break previously working code. Maybe justly so, but I'm sure users will find this confusing.

Tony's reply to my concern: 

> Regarding the last point, I don't think that breakage would occur in practice because forward-declaring a Swift @objc class already shouldn't be expected to work. As an example, let's say Swift module A defines an @objc class, and Clang module B only forward declares that class and has some API that references it. Even if Swift module C that imports both A and B, the APIs in B that refer to A won't work, so anyone trying to forward-declare a Swift class in Obj-C is already putting their build in a fragile and unreliable state.

I'm not trying to advocate for forward declaring Swift types as a supported practice, but I do want to call out that I think this breakage can occur in practice. Firstly, Clang module B is usable by Objective-C clients without issue, and anecdotally that is where we run into most forward declaration issues (ie. only Objective-C clients used the interface before and had no problems with forward declarations, then a Swift client arrives and has problems). Even if clients shouldn't expect it to work, unfortunately it has worked for a long time, so I don't think its unlikely for clients to have working code on the ground doing this. 

If Swift module C imports only Clang module B, after this patch the APIs _will_ work, since the compiler thinks the forward declarations refer to Clang definitions. It will only break once Swift module A is imported. So it is possible a Swift client starts using a Clang module that forward declares Swift types, and everything works for some time until they import the complete definition of that type. That said, I'm not overly worried about this since I think its relatively uncommon and easy to fix.

I'm comfortable with both these caveats, and I don't see a better solution, but I wanted to make them explicit for others so that there are no surprises down the line.

cc: @allevato @beccadax @drodriguez @plotfi

plotfi commented on lib/AST/Module.cpp:
```diff
@@ -1055,6 +1056,62 @@ void ModuleDecl::getTopLevelDeclsWhereAttributesMatch(
+
+  for (Decl *decl : topLevelObjCExposedDeclsInFileUnit) {
+    if (ValueDecl *VD = dyn_cast<ValueDecl>(decl)) {
+      if (VD->hasName()) {
```
nit: avoid scope nesting by doing something like this:

```
ValueDecl *VD = dyn_cast<ValueDecl>(decl);
if (!VD || !VD->hasName())
  continue;
```

Don't worry about it in this instance though. I think this is fine.

plotfi commented on lib/AST/Module.cpp:
```diff
@@ -1055,6 +1056,62 @@ void ModuleDecl::getTopLevelDeclsWhereAttributesMatch(
+    }
+  }
+
+  for (Decl *decl : topLevelObjCExposedDeclsInFileUnit) {
```
Could `Decl *decl` be `const Decl *decl` here?

plotfi commented:
I looked over the patch once myself and once with Nuri. Overall this PR looks good to me, and the set of included test cases looks very goods.

However I am not the code owner, could Nuri please get a re-review for this PR please? Thanks 

 @beccadax @drodriguez @zoecarver @allevato @hyp

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
Hi @allevato @beccadax,

With the Swift 5.9 branch cut quickly approaching on March 20th, we'd like to merge this in time for the release. The implementation hasn't changed for a month and if no new concerns arise, I intend on merging in the next week. If concerns crop up afterwards, I'll be available to fix them.

That said, @DougGregor it seems you are the 5.9 release manager for the Swift compiler. If you have the time, I'd appreciate a review; an approval from the Apple side before merging would be preferred.

zoecarver reviewed (approved):
I did a quick review (not very in-depth) and this looks good to me. I think it would still be good to get an in depth review from Becca, but I understand that she might not have time. 

Please make sure to run the source compatibility test suite before landing this.

NuriAmari commented:
@swift-ci Please test

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@swift-ci Please test source compatibility

NuriAmari commented:
@swift-ci Please test source compatibility
